File: Javascript/Jeu_min_max.js
// # Importations
const fs = require("fs");
const path = require("path");

// # Données
const CHEMIN_DOSSIER = __dirname;
const FICHIER_OPTIONS = path.join(CHEMIN_DOSSIER, "minmax_options.txt");

// # Fonctions utilitaires
// ## Lecture d'une ligne synchronisée
/**
	Lit une ligne sur STDIN de façon synchrone
	@returns ligne lue (sans « \n »)
*/
function readLineSync() {
	const tampon = Buffer.alloc(1);
	let ligne = "";
	while (true) {
		const octetsLus = fs.readSync(0, tampon, 0, 1, null);
		if (octetsLus === 0) break;
		const caractere = tampon.toString();
		if (caractere === "\n") break;
		ligne += caractere;
	}
	return ligne;
}

// ## Lecture d'un entier utilisateur sécurisé
/**
	Lit un entier entre min et max inclus, boucle jusqu'à validité
	@message prompt
	@min borne basse
	@max borne haute
*/
function lireEntier(message, min, max) {
	while (true) {
		process.stdout.write(message);
		const entree = readLineSync().trim();
		if (/^\d+$/.test(entree)) {
			const nombre = Number(entree);
			if (nombre >= min && nombre <= max) return nombre;
		}
		console.log(`\nVeuillez entrer un entier entre ${min} et ${max}.`);
	}
}

// ## Chargement des options sauvegardées
/**
	Charge le fichier d’options ou renvoie les valeurs par défaut
*/
function chargerOptions() {
	if (fs.existsSync(FICHIER_OPTIONS)) {
		const donnees = fs
			.readFileSync(FICHIER_OPTIONS, "utf8")
			.split(",")
			.map(Number);
		if (donnees.length === 3) return donnees;
	}
	return [1, 100, 5];
}

// ## Sauvegarde des options
/**
	Enregistre les options actuelles
*/
function sauvegarderOptions(min, max, tours) {
	fs.writeFileSync(FICHIER_OPTIONS, [min, max, tours].join(","));
}

// # Fonctions principales
/**
	Affiche le menu principal
*/
function afficherMenuPrincipal() {
	console.log("\n#### Menu ####\n1 : Jouer\n2 : Options\n3 : Quitter");
}

/**
	Affiche le menu d’options
*/
function afficherMenuOptions(min, max, tours) {
	console.log(
		`\n#### Menu options ####\n1 : Choisir les limites (actuellement : ${min} - ${max})\n2 : Nombre de tours max (actuellement : ${tours})`
	);
}

/**
	Lance une partie « plus-grand / plus-petit »
*/
function jouerPartie(min, max, tours) {
	const cible = Math.floor(Math.random() * (max - min + 1) + min);
	console.log(`\nTrouvez entre ${min} et ${max} en ${tours} tours`);
	for (let tourNumero = 1; tourNumero <= tours; tourNumero++) {
		const choix = lireEntier(`Tour ${tourNumero} : `, min, max);
		if (choix === cible) {
			console.log("\nGagné !");
			return;
		}
		console.log(choix > cible ? "-" : "+");
	}
	console.log(`\nPerdu ! La réppnse était ${cible}.`);
}

/**
	Gère le menu Options puis sauvegarde
*/
function gererOptions(min, max, tours) {
	afficherMenuOptions(min, max, tours);
	const choixOptions = lireEntier("? = ", 1, 2);
	if (choixOptions === 1) {
		min = lireEntier("\nMin = ", 1, max);
		max = lireEntier("\nMax = ", min, 10000);
	} else if (choixOptions === 2) {
		tours = lireEntier("\nTours max = ", 1, 100);
	}
	sauvegarderOptions(min, max, tours);
	return [min, max, tours];
}

// # Main
/**
	Point d’entrée principal
*/
function main() {
	let [min, max, tours] = chargerOptions();
	while (true) {
		afficherMenuPrincipal();
		const choixMenu = lireEntier("? = ", 1, 3);
		if (choixMenu === 1) {
			jouerPartie(min, max, tours);
		} else if (choixMenu === 2) {
			[min, max, tours] = gererOptions(min, max, tours);
		} else if (choixMenu === 3) {
			break;
		}
	}
}

// # Lancement du programme
main();

File: C#/src/Jeu_min_max.cs
// Importations
using System;
using System.IO;

class Program
{
	// Données
	private static readonly string cheminDossier = Directory.GetCurrentDirectory();

	private static readonly string fichierOptions = Path.Combine(cheminDossier, "data", "minmax_options.txt");

	// Création automatique du dossier data
	static Program()
	{
		string dossierData = Path.Combine(cheminDossier, "data");
		if (!Directory.Exists(dossierData))
		{
			Directory.CreateDirectory(dossierData);
		}
	}

	// Fonctions utilitaires
	// ## Lecture d'un entier utilisateur sécurisé
	/**
		Lecture d'un entier utilisateur sécurisé
		Boucle jusqu'à obtenir un entier valide entre min et max.
		@message invite à l'utilisateur
		@min borne inférieure
		@max borne supérieure
		@return entier validé
	*/
	static int LireEntier(string message, int min, int max)
	{
		while (true)
		{
			Console.Write(message);
			string entree = Console.ReadLine().Trim();
			if (int.TryParse(entree, out int nombre) && nombre >= min && nombre <= max) return nombre;
			Console.WriteLine($"\nVeuillez entrer un entier entre {min} et {max}.");
		}
	}

	// Fonctions utilitaires principales
	// ## Chargement et sauvegarde des options
	/**
		Charge les options ou renvoie les valeurs par défaut.
		@return tuple (min,max,tours)
	*/
	static (int min, int max, int tours) ChargerOptions()
	{
		if (File.Exists(fichierOptions))
		{
			string[] donnees = File.ReadAllText(fichierOptions).Split(',');
			if (donnees.Length == 3 &&
				int.TryParse(donnees[0], out int min) &&
				int.TryParse(donnees[1], out int max) &&
				int.TryParse(donnees[2], out int tours)) return (min, max, tours);
		}
		return (1, 100, 5);
	}

	/**
		Sauvegarde les options actuelles.
	*/
	static void SauvegarderOptions(int min, int max, int tours)
	{
		File.WriteAllText(fichierOptions, $"{min},{max},{tours}");
	}

	// Fonctions principales
	/**
		Affiche le menu principal.
	*/
	static void AfficherMenuPrincipal()
	{
		Console.WriteLine("\n#### Menu ####\n1 : Jouer\n2 : Options\n3 : Quitter");
	}

	/**
		Affiche le menu d'options.
	*/
	static void AfficherMenuOptions(int min, int max, int tours)
	{
		Console.WriteLine($"\n#### Menu options ####\n1 : Choisir les limites (actuellement : {min} - {max})\n2 : Nombre de tours max (actuellement : {tours})");
	}

	/**
		Joue une partie, indique si plus grand/petit.
	*/
	static void JouerPartie(int min, int max, int tours)
	{
		Random generateur = new Random();
		int cible = generateur.Next(min, max + 1);
		Console.WriteLine($"\nTrouvez entre {min} et {max} en {tours} tours");
		for (int essaiNumero = 1; essaiNumero <= tours; essaiNumero++)
		{
			int choix = LireEntier($"Tour {essaiNumero} : ", min, max);
			if (choix == cible) { Console.WriteLine("\nGagné !"); return; }
			Console.WriteLine(choix > cible ? "-" : "+");
		}
		Console.WriteLine($"\nPerdu ! La réppnse était {cible}.");
	}

	/**
		Gère les options et les sauvegarde si modifiées.
		@return tuple (min,max,tours)
	*/
	static (int min, int max, int tours) GererOptions(int min, int max, int tours)
	{
		AfficherMenuOptions(min, max, tours);
		int choixOptions = LireEntier("? = ", 1, 2);
		if (choixOptions == 1)
		{
			min = LireEntier("\nMin = ", 1, max);
			max = LireEntier("\nMax = ", min, 10000);
		}
		else if (choixOptions == 2) { tours = LireEntier("\nTours max = ", 1, 100); }
		SauvegarderOptions(min, max, tours);
		return (min, max, tours);
	}

	// Main
	/**
		Boucle principale : menu, options, jeu.
	*/
	static void main()
	{
		var (min, max, tours) = ChargerOptions();
		while (true)
		{
			AfficherMenuPrincipal();
			switch (LireEntier("? = ", 1, 3))
			{
				case 1: JouerPartie(min, max, tours); break;
				case 2: (min, max, tours) = GererOptions(min, max, tours); break;
				case 3: return;
			}
		}
	}

	// Lancement du programme
	static void Main(string[] args) => main();
}
